src/core/compiler.jl

```diff
@@ -7,124 +7,148 @@ using Base.Meta: parse
 macro VarName(ex::Union{Expr, Symbol})
   # Usage: @VarName x[1,2][1+5][45][3]
   #    return: (:x,[1,2],6,45,3)
-  s = string(gensym())
-  if isa(ex, Symbol)
-    ex_str = string(ex)
-    return :(Symbol($ex_str), (), Symbol($s))
-  elseif ex.head == :ref
-    _2 = ex
-    _1 = ""
-    while _2.head == :ref
-      if length(_2.args) > 2
-        _1 = "[" * foldl( (x,y)-> "$x, $y", map(string, _2.args[2:end])) * "], $_1"
-      else
-        _1 = "[" * string(_2.args[2]) * "], $_1"
-      end
-      _2   = _2.args[1]
-      isa(_2, Symbol) && (_1 = ":($_2)" * ", ($_1), Symbol(\"$s\")"; break)
+    s = string(gensym())
+    if isa(ex, Symbol)
+        ex_str = string(ex)
+        return :(Symbol($ex_str), (), Symbol($s))
+    elseif ex.head == :ref
+        _2 = ex
+        _1 = ""
+        while _2.head == :ref
+            if length(_2.args) > 2
+                _1 = "[" * foldl( (x,y)-> "$x, $y", map(string, _2.args[2:end])) * "], $_1"
+            else
+                _1 = "[" * string(_2.args[2]) * "], $_1"
+            end
+            _2 = _2.args[1]
+            isa(_2, Symbol) && (_1 = ":($_2)" * ", ($_1), Symbol(\"$s\")"; break)
+        end
+        return esc(parse(_1))
+    else
+        @error "VarName: Mis-formed variable name $(ex)!"
+        return :()    
     end
-    return esc(parse(_1))
-  else
-    error("VarName: Mis-formed variable name $(ex)!")
-  end
 end
 
+function generate_observe(left, right)
+    obsexpr = esc( quote
+            _lp += Turing.observe(
+                                  sampler,
+                                  $(right),   # Distribution
+                                  $(left),    # Data point
+                                  vi
+                                 )
+        end )
+    return obsexpr
+end
+
+
 """
-    var_name ~ Distribution()
+    macro: @~ var Distribution()
 
-Tilde notation `~` can be used to specifiy *a variable follows a distributions*.
+Tilde notation macro. This macro constructs Turing.observe or 
+Turing.assume calls depending on the left-hand argument.
+Note that the macro is interconnected with the @model macro and 
+assumes that a `compiler` struct is available.
 
-If `var_name` is an un-defined variable or a container (e.g. Vector or Matrix), this variable will be treated as model parameter; otherwise if `var_name` is defined, this variable will be treated as data.
+Example:
+```julia
+@~ x Normal()
+```
 """
 macro ~(left, right)
```

Comment:
It will be invoked if the left-hand side is a constant, e.g.
```julia
1 ~ Bernoulli(x / 2)
```
which doesn't make much sense to me, but we have it in the tests. See `test/compiler.jl/observe.jl`.

@yebai or @xukai92 what is the purpose for this test?

Reply:
This feature is only useful when a user writes a model like

```julia
@model gdemo(x) = begin
  s ~ InverseGamma(2,3)
  m ~ Normal(0,sqrt(s))
  1 ~ Normal(m, sqrt(s))
  2 ~ Normal(m, sqrt(s))
  return s, m
end
```
Here all the observations are provided as constants.